Bridge/src/zeta.cpp

```diff
@@ -0,0 +1,116 @@
+#include "zeta.hpp"
+
+#include <chrono>
+#include <iostream>
+
+#include "debug.hpp"
+
+ZetaRfRadio::ZetaRfRadio()
+{
+    std::cout << "Starting Zeta Radio..." << std::endl;
+
+    if (!this->zeta_.beginWithPacketLengthOf(PACKET_LENGTH))
+        throw "ZetaRf begin failed. Check wiring?";
+
+    if (!this->zeta_.startListeningSinglePacketOnChannel(ZetaRfRadio::ZETA_CHANNEL))
+        throw "ZetaRf startListening failed.";
+
+    debug << "Starting ZetaRf worker thread..." << std::endl;
+    this->worker_ = std::thread(&ZetaRfRadio::rx_tx_loop, this);
+
+    std::cout << "ZetaRf init done." << std::endl;
+}
+
+ZetaRfRadio::~ZetaRfRadio()
+{
+    if (this->worker_.joinable())
+    {
+        this->should_worker_join_ = true;
+        this->worker_.join();
+    }
+}
+
+void ZetaRfRadio::set_on_received(std::function<void(Frame)> callback)
+{
+    this->on_receive_ = callback;
+}
+
+void ZetaRfRadio::send_packets(const std::vector<Frame> frames)
+{
+    this->send_queue_.push(frames);
+}
+
+void ZetaRfRadio::rx_tx_loop()
+{
+    while (!this->should_worker_join_)
+    {
+        // Receive packet
+        this->process_zeta_events();
+        if (this->zeta_.hasDataAvailable())
+            this->read_packet();
+
+        // Transmit next queued packet.
+        // We could send all queue packets, but we don't want to overwhelm the
+        // receiver's rx fifo if it is also transmitting.
+        if (auto packet = this->send_queue_.pop())
+        {
+            debug << "Sending packet: " << &*packet << std::endl;
+            this->transmit_packet(*packet);
+        }
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(2));
+    }
+    debug << "ZetaRf worker thread exiting..." << std::endl;
+}
+
+void ZetaRfRadio::process_zeta_events()
+{
+    const auto event = this->zeta_.checkForEvent();
+    if (event & ZetaRf::Event::DeviceBusy)
```

Comment:
Would a switch case statement be better here instead? Are we only expecting these two events btw?

Reply:
A switch won't work because `event` is an integer, not an enum, and it's possible to receive more than one events at once.

These are the only two events we're handling, but there's heaps more! https://github.com/GIPdA/ZetaRF/blob/v2/src/zetarf_radio.hpp